Add initial draft step to GitHub release

Luehrsen opened the pull request:
This PR moves the actual publishing of the GitHub release to after the release assets have been uploaded.

This is important because Github Action triggers a release on publish, so it is possible that release assets might not be visible to the action, as they are yet to be uploaded.

see #https://github.com/release-it/release-it/issues/495

webpro commented:
Thanks, @Luehrsen! Looks good. Could you please add coverage for this in the tests?

Luehrsen commented:
Will do...

Luehrsen commented:
As far as I can see 'github should release and upload assets' https://github.com/release-it/release-it/blob/master/test/github.js#L35 already covers this code. 

The fail-case would be, that a release is drafted although draft is not `true`. The tests already test if a tag_name is assigned. The tag_name itself is only assigned on published releases, as drafts do not yet have a valid tag. 

As for some of the other tests that currently fail: I have no idea how to fix them. AFAIK the also fail on master.

webpro commented:
* You can see what parts of the code aren't covered yet with `npm run coverage` (results in the `coverage` dir).
* The tests should run without problems: https://travis-ci.org/release-it/release-it/builds
* I can see the failing tests from this branch here: https://travis-ci.org/release-it/release-it/jobs/524915444#L634. The first error makes sense: the `draft` prop is now `true` (not `false` anymore) after your changes. And so on. Or are you referring to other failing tests?

Luehrsen commented:
Ah, this is my first time working with ava (and TDD is not part of my usual workload), so sorry for the confusion. 

I have started to implement the tests, but I've hit somewhat of a roadblock: updateRelease on L184 returns a rejected promise, that leads the following code to fail. Any idea on how to get that to work? 

Thanks!

Luehrsen commented:
Found my issue: I need a stub. Working on it now...

Luehrsen commented:
I still have quite some issues while debugging the tests. I'm not seeing the same fails locally as travis is seeing. For me, all git related tests fail.

```
  should run tasks without throwing errors

  /Projekte/wp_development/projects/release-it/tmp/9953082f-daad-4cd0-b2f1-3496b96ad725/lib/shell.js:44

  Rejected promise returned by test. Reason:

  Error {
    message: `No refs in common and none specified; doing nothing.␊
    Perhaps you should specify a branch such as 'master'.␊

    fatal: the remote end hung up unexpectedly␊
    error: failed to push some refs to '/Projekte/wp_development/projects/release-it/tmp/fe5eda00-09dc-4e9b-b04f-a7e280d3fc70'␊
    `,
  }

```

I believe there is some issue in the git config on my system, but I can't figure out what. (And I haven't had any issues up until now...)

Luehrsen commented:
Found the issue: My local git config defaulted to 'matching' as push default, but this setup expects 'simple'. 

See: https://git-scm.com/docs/git-config#Documentation/git-config.txt-pushdefault

webpro commented on test/github.js:
```diff
@@ -58,12 +58,23 @@ test('should release and upload assets', async t => {
 
   t.is(releaseResult.tag_name, 'v' + version);
   t.is(releaseResult.name, 'Release ' + version);
-  t.is(githubRequestStub.callCount, 1);
```
We can keep this assertion.

webpro commented on lib/tasks.js:
```diff
@@ -130,8 +130,12 @@ const runTasks = async (opts, injected = {}) => {
+        const release = () =>
+          ghRelease()
+            .then(() => ghUploadAssets())
+            .then(() => ghPublishRelease());
```
Please also update the flow for `isInteractive === false` below. This will then fix the `draft: true` assertions in `test/tasks.js` ;-)

The flows are slightly different. After prompted to release to GitHub, there should be a single `task` to complete. But in non-interactive mode, it's nice to separate the actions to better indicate the progress.

webpro commented on lib/tasks.js:
```diff
@@ -130,8 +130,12 @@ const runTasks = async (opts, injected = {}) => {
       github.release && github.releaseNotes && logPreview(log, 'release notes', await ghClient.getNotes());
       const ghRelease = () => ghClient.release({ version, isPreRelease, changelog });
       const ghUploadAssets = () => ghClient.uploadAssets();
+      const ghPublishRelease = () => ghClient.maybePublishRelease();
```
Maybe with `maybePublishRelease` it makes sense to rename `release` to `draftRelease`?

Luehrsen commented:
Hey @webpro, I hope you had a great weekend. 

I've renamed the release function and added the non-interactive mode. But now I've got some failed tests which I believe have to do with the order of calls being made. Maybe even a race condition? I really don't understand what is going on there, maybe you can have a look.

webpro commented:
The situation is that we are now going from two to three calls to GitHub (or the stub in test runs). The bug in the PR is that `maybePublishRelease` overwrites the url (`this.releaseUrl = html_url`) from `draftRelease` while it's `undefined`. So we can just remove this line from the `draftRelease` as it was already set.

Please find attached the patch file. Remove the `.txt` extension and apply with `patch -p0 -i pr512.patch`:

[pr512.patch.txt](https://github.com/release-it/release-it/files/3126402/pr512.patch.txt)

I've also added a few assertions for both related calls. Hopefully this clears things up a little. If not, please do not hesitate to let me know :)

Luehrsen commented:
The tests are all green, but I've actually found an issue: 

The release_url that is getting printed to the command line is wrong. It is always the draft URL, no matter if it is released or not. So we actually need to overwrite the releaseUrl. (Or find another method to overwrite that.)

webpro commented:
Are the url's different? What is printed? It should be `https://github.com/${owner}/${repoName}/releases/tag/{tag}`

Luehrsen commented:
The format is correct, the tag is wrong. 

This is what gets printed: 
`🔗 https://github.com/luehrsenheinrich/release-it-test/releases/tag/untagged-1e51fa1f91e3c1b00d61`

This is what it should be:
`https://github.com/luehrsenheinrich/release-it-test/releases/tag/1.0.12`

webpro commented:
Ah I see. Then we need to inspect the response from the `repos.updateRelease` call and hopefully the url is in there...?

You can use `DEBUG=release-it:* release-it` to have this response logged.

Luehrsen commented on lib/github.js:
```diff
@@ -183,9 +183,7 @@ class GitHub extends Release {
         debug(options);
         const response = await this.client.repos.updateRelease(options);
         debug(response.data);
-        const { html_url } = response.data;
```
This is where the updated release URL would come out, this is why this override exists...

Luehrsen replied:
*existed. This got removed by the patch. To we need that, but we also need that to play nice with the stubs. 🤔

webpro replied:
OK now I get it. We can send the `tag_name` again with the `options` in the `repos.updateRelease` (so it will also come back with the stub response) and restore the line I removed here. Apologies for the confusion!

Luehrsen replied:
Well, this should be it...

webpro commented:
Thanks a lot, great work! I've squashed your work into [v10.4.5](https://github.com/release-it/release-it/releases/tag/10.4.5).
